### PR TITLE
Require build class parent

### PR DESCRIPTION
**Pull request by temyurchenko**
A part of getting rid of non-Module roots, see #2536

### CODE REVIEW

File: tests/brain/test_brain.py [hunk 1]
```diff
@@ -1705,7 +1705,8 @@ def test_infer_dict_from_keys() -> None:
     )
     for node in bad_nodes:
         with pytest.raises(InferenceError):
-            next(node.infer())
+            if isinstance(next(node.infer()), util.UninferableBase):
```
Comment (DanielNoord) [inline tests/brain/test_brain.py:1708]: Why is this necessary?

File: astroid/nodes/node_classes.py [hunk 2]
```diff
@@ -2061,7 +2061,14 @@ def __init__(
         :param end_col_offset: The end column this node appears on in the
             source code. Note: This is after the last symbol.
         """
-        self.value: Any = value
+        if getattr(value, "__name__", None) == "__doc__":
+            raise AssertionError(  # pragma: no cover
```
Comment (DanielNoord) [inline astroid/nodes/node_classes.py:2065]: I don't think this belongs in this PR right?

File: astroid/nodes/scoped_nodes/scoped_nodes.py [hunk 3]
```diff
@@ -178,6 +178,13 @@ def function_to_method(n, klass):
     return n
 
 
+def attach_to_parent(node: NodeNG, name: str, parent: NodeNG):
```
Comment (DanielNoord) [inline astroid/nodes/scoped_nodes/scoped_nodes.py:181]:
```suggestion
def _attach_to_parent(node: NodeNG, name: str, parent: NodeNG):
```

File: astroid/nodes/scoped_nodes/scoped_nodes.py [hunk 4]
```diff
@@ -178,6 +178,13 @@ def function_to_method(n, klass):
...
+    frame = parent.frame()
+    frame.set_local(name, node)
+    if frame is parent:
+        frame._append_node(node)
```
Comment (DanielNoord) [inline astroid/nodes/scoped_nodes/scoped_nodes.py:185]: This feels like a crucial change that we don't have a test for. Should we add one?

**Review by DanielNoord — COMMENTED** [review]
Awesome!

File: astroid/nodes/node_classes.py (hunk 2 above)
Comment (temyurchenko) [reply astroid/nodes/node_classes.py:2065]: The first three commits in this PR will be merged in the previous PRs. Unfortunately, they are a necessity. The issue that these commits fix are uncovered by the last commit, so without them there are going to be errors.

File: tests/brain/test_brain.py (hunk 1 above)
Comment (temyurchenko) [reply tests/brain/test_brain.py:1708]: For this one, I don't know exactly, but my intuition was the following. Before, when we had `Unknown` for a parent, when inference encountered Unknown, it would raise (`InferenceError`). Now, there is no `Unknown`, but the calls still aren't correct. So, we just fail to infer and yield `Uninferable`. 

However, both responses kind of seem to make sense, so I opted to allow either of the responses (`InferenceError` or `Uninferable`) to pass the tests.

Comment (DanielNoord) [reply tests/brain/test_brain.py:1708]: Can we be certain that it is either of the two? So, just assert that it is always `Uniferable` or always the exception? If so, I'd prefer just testing for that

Comment (temyurchenko) [reply tests/brain/test_brain.py:1708]: Yes, right now it's always `Uninferable`. I'll change to test specifically for that.

File: astroid/nodes/scoped_nodes/scoped_nodes.py (hunk 4 above)
Comment (temyurchenko) [reply astroid/nodes/scoped_nodes/scoped_nodes.py:185]: So, usually the classes get added to the body of the parent in the positinit constructor. There is only one exception to my knowledge: `raw_building._add_dunder_class`. It was added solely to support some `pyqt` inference and it is creating (and attaching) adhoc classes for the purposes of being later detected by the transform process. 

That situation is already tested by the three qt tests (which fail without this change). In other situations it's useless, afaics.

Overall, the process described above seems very perverse, the "dunder" class shouldn't be recreated within a function each time. It is its own class that is defined outside. We should rather consult the type annotation in the transform process. However, it is a question for another PR.

File: astroid/nodes/scoped_nodes/scoped_nodes.py (hunk 3 above)
Comment (temyurchenko) [reply astroid/nodes/scoped_nodes/scoped_nodes.py]: fixed

File: tests/brain/test_brain.py (hunk 1 above)
Comment (temyurchenko) [reply tests/brain/test_brain.py]: fixed

Comment (temyurchenko) [reply tests/brain/test_brain.py]: Ah, one test fails under `pypy` with `InferenceError`. I'd rather go with the initial `pytest.raises` clause then

**Review by DanielNoord — APPROVED** [review]
